src/widget/wtrackmenu.h

```diff
@@ -0,0 +1,255 @@
+    void setTrackPointerList(TrackPointerList trackPointerList);
+    TrackPointerList getTrackPointerList();
+
+    bool modelHasCapabilities(TrackModel::CapabilitiesFlags capabilities) const;
+    bool optionIsEnabled(Filter flag);
+
+    void addSelectionToPlaylist(int iPlaylistId);
+    void updateSelectionCrates(QWidget* pWidget);
+
+    void showTrackInfo(QModelIndex index);
+    void showDlgTagFetcher(QModelIndex index);
+
+    void addToAutoDJ(PlaylistDAO::AutoDJSendLoc loc);
+
+    void teardownActions();
+
+
+    void loadSelectionToGroup(QString group, bool play = false);
+    void clearTrackSelection();
+
+    ControlProxy* m_pNumSamplers{};
+    ControlProxy* m_pNumDecks{};
+    ControlProxy* m_pNumPreviewDecks{};
+
+    // The selected tracks for which the context menu is created
+    TrackIdList m_pTrackIdList;
+
+    // Context menu machinery
+    QMenu *m_pLoadToMenu{};
+    QMenu *m_pDeckMenu{};
+    QMenu *m_pSamplerMenu{};
+
+    QMenu *m_pPlaylistMenu{};
+    QMenu *m_pCrateMenu{};
+    QMenu *m_pMetadataMenu{};
+    QMenu *m_pMetadataUpdateExternalCollectionsMenu{};
+    QMenu *m_pClearMetadataMenu{};
+    QMenu *m_pBPMMenu{};
+    QMenu *m_pColorMenu{};
+
+
+    WCoverArtMenu* m_pCoverMenu{};
+
+    // Reload Track Metadata Action:
+    QAction *m_pImportMetadataFromFileAct{};
+    QAction *m_pImportMetadataFromMusicBrainzAct{};
+
+    // Save Track Metadata Action:
+    QAction *m_pExportMetadataAct{};
+
+    // Load Track to PreviewDeck
+    QAction* m_pAddToPreviewDeck{};
+
+    // Send to Auto-DJ Action
+    QAction *m_pAutoDJBottomAct{};
+    QAction *m_pAutoDJTopAct{};
+    QAction *m_pAutoDJReplaceAct{};
+
+    // Remove from table
+    QAction *m_pRemoveAct{};
+    QAction *m_pRemovePlaylistAct{};
+    QAction *m_pRemoveCrateAct{};
+    QAction *m_pHideAct{};
+    QAction *m_pUnhideAct{};
+    QAction *m_pPurgeAct{};
+
+    // Show track-editor action
+    QAction *m_pPropertiesAct{};
+    QAction *m_pFileBrowserAct{};
+
+    // BPM feature
+    QAction *m_pBpmLockAction{};
+    QAction *m_pBpmUnlockAction{};
+    QAction *m_pBpmDoubleAction{};
+    QAction *m_pBpmHalveAction{};
+    QAction *m_pBpmTwoThirdsAction{};
+    QAction *m_pBpmThreeFourthsAction{};
+    QAction *m_pBpmFourThirdsAction{};
+    QAction *m_pBpmThreeHalvesAction{};
+
+    // Track color
+    WColorPickerAction *m_pColorPickerAction{};
+
+    // Clear track metadata actions
+    QAction* m_pClearBeatsAction{};
+    QAction* m_pClearPlayCountAction{};
+    QAction* m_pClearMainCueAction{};
+    QAction* m_pClearHotCuesAction{};
+    QAction* m_pClearIntroCueAction{};
+    QAction* m_pClearOutroCueAction{};
+    QAction* m_pClearLoopAction{};
+    QAction* m_pClearWaveformAction{};
+    QAction* m_pClearKeyAction{};
+    QAction* m_pClearReplayGainAction{};
+    QAction* m_pClearAllMetadataAction{};
+
+
+    const UserSettingsPointer m_pConfig;
+
+    TrackCollectionManager* m_pTrackCollectionManager;
+    TrackPointerList m_pTrackPointerList;
```

Comment:
I will resort to using getter functions and only store the track pointers.
But the track indices still need to be stored separately.

The client will only (ideally) supply only the `TrackIdList` or `QModelIndexList`, member variables have to be created for both of these possibilities.

Reply:
I was looking for ways to remove the QModelIndexes in WTrackMenu, but yes I think you are right that having members for both the Tracks and the QModelIndexes is required because some features need to pass a QModelIndex to the TrackModel* and the next/previous buttons in DlgTrackInfo and DlgTagFetcher require the QModelIndexes.